# Verzweigungen
Verzweigungen gehören neben den einfachen Anweisungen und den Wiederholungen zu den drei Grundaktionen, aus denen Algorithmen bestehen.

Eine Verzweigung prüft immer eine Bedingung. Ist die Bedingung wahr (<code>True</code>), wird ein Anweisungsblock ausgeführt. Ist die Bedingung Falsch (<code>False</code>), wird entweder nichts oder ein anderer Teil ausgeführt.

In [ ]:
a = int(input("Bitte Wert für a eingeben: "))
if (a > 5):
    print("a ist größer als 5.")
    
b = int(input("Bitte Wert für b eingeben: "))
if (b > 5):
    print("b ist größer als 5.")

Das <code>if</code> Schlüsselwort prüft eine Bedingung und führt, sofern diese Wahr ist, den zugehörigen Codeblock aus.

Damit klar ist, welche Anweisungen zu einem Codeblock gehören, werden diese gleichmäßig eingerückt. Sobald eine Anweisung nach dem if-codeblock wieder ganz am linken Zeilenrand beginnt, ist klar dass sie nicht mehr zum If statement gehört.

In [ ]:
a = int(input("Bitte Wert für a eingeben: "))
if (a == 3):
    print("a ist gleich drei!")
    print("Dieser Text wird nur angezeigt, wenn a genau drei ist.")
print("Dieser Text wird immer angezeigt, da die Anweisung nicht mehr zum if-codeblock gehört.")

## Wenn die Bedingung nicht erfüllt ist
Für den Fall, dass die Bedingung nicht erfüllt ist, wird in den obigen Beispielen die if Anweisung mit dem gesamten dazugehörigen Codeblock übersprungen.

Soll ein Codeblock hinzugefügt werden, der **nur dann** ausgeführt wird, wenn eine Bedingung **nicht** Wahr ist, so muss das Schlüsselwort <code>else</code> verwenden werden:

In [ ]:
a = int(input("Bitte Wert für a eingeben: "))
if (a == 3):
    print("a ist genau 3.")
else:
    print("a ist nicht gleich 3.")

b = int(input("Bitte Wert für b eingeben: "))
if (b == 3):
    print("b ist genau 3.")
else:
    print("b ist nicht gleich 3.")

# Mehrere Bedingungen prüfen

Für den Fall, dass mehrere Bedingungen geprüft werden sollen, gibt es zusätzlich das Schlüsselwort <code>elif</code>. Die Anweisungen die zu einem <code>elif</code> Zweig gehören (durch entsprechende Einrückung gekennzeichnet), werden nur dann ausgeführt, wenn die Bedingung nach dem zugehörigen <code>if</code> Statement nicht wahr ist aber die Bedingung nach dem <code>elif</code> Statement wahr ist.

Ein <code>elif</code> Block kann immer nur nach einem <code>if</code> oder einem anderen <code>elif</code> Block stehen. Natürlich wäre es auch möglich, statt des <code>elif</code> einfach ein neues <code>if</code> zu schreiben. Aber das wäre sehr ineffizient, da stets beide Ausdrücke ausgewertet werden müssen, wohingegen die <code>elif</code> Lösung nur ausgewertet wird, wenn die <code>if</code> Anweisung erfolglos war.

In [ ]:
a = 3
b = 5

if (a == 4):
    print("a ist genau 4!")
elif (b == 5):
    print("a ist zwar nicht gleich 4 aber dafür ist b gleich 5!")
else:
    print("a ist nicht gleich 4 und b ist nicht gleich 5!")
    
#-----------------------------------------------------------------

wert = int(input("Bitte eine Zahl eingeben: "))

if (wert < 3):
    print("Wert ist echt kleiner als 3.")
elif (wert >=3) and (wert < 9):
    print("Wert ist im Bereich 3 (inklusive) bis 9 (exklusive).")
else:
    print("Wert ist größer oder gleich 9.")

## Alternative Schreibweisen in Python für "leichte" Verzweigungen

Es soll im Folgenden noch kurz auf alternative Schreibweisen für Verzweigungen eingegangen werden. Damit sollen Sie in den Stand versetzt werden, dass Sie bestehenden Python Code lesen können. Speziell zu Beginn wird empfohlen, die oben beschriebene Schreibweise zu nutzen, da sie übersichtlicher ist und etwas weniger fehleranfällig.


In [ ]:
if (x == 1):
    a = 1
else:
    a = 2

Dabei handelt es sich um einen einfachen Ausdruck, der der Variable a unterschiedliche Werte zuweist für unterschiedliche x Werte. Dies ist eine recht aufwendige Schreibweise. Daher gibt es folgende Kurzschreibweise:

In [ ]:
a = (1 if (x == 1) else 2)

Die Klammern sind in diesem Fall nicht notwendig, erhöhen aber die Lesbarkeit ungemein. Diese Schreibweise wird auch **Bedingte Ausdrücke** genannt.

Die allgemeine Formulierung eines Bedingten Ausdrucks lautet:
**A if Bedingung else B**.
A und B sind dabei Anweisungen und nicht nur einfache Zuweisungen. Dadurch können auch folgende <code>print()</code> Anweisungen geschrieben werden:

In [ ]:
x = 5
print("x hat den Wert 1" if (x == 1) else "x ist ungleich 1")

Erneut sind die Klammern um <code> x == 1 </code> nicht notwendig, aber guter Programmierstil.

# Frage nach größten Integer und Float Zahlen

Die größten Integer und Float Zahlen können folgendermaßen ermittelt werden. Dabei wird eine Bibliothek eingebunden. Näheres dazu erfahren wir in späteren Vorlesungen.

In [1]:
import sys #das ist die Bibliothek
print("Max Integer: ", sys.maxsize)
print("Max Float: ", sys.float_info.max)

Max Integer:  9223372036854775807
Max Float:  1.7976931348623157e+308


Hierbei ist anzumerken, dass die E-Schreibweise genutzt wurde. Der maximale Float Wert ist mathematisch $$1.7976931348623157 \cdot 10^{308}$$

Die E-Schreibweise führt nach dem "e" den Exponenten zur Basis 10 auf. Für Float Zahlen ist die Eingabe mit der "E-Schreibweise" auch möglich. Vor den Exponenten kann natürlich auch ein "-" geschrieben werden um zum Beispiel diese Zahl zu spezifizieren:
$$1.7976931348623157 \cdot 10^{-308}$$